packages/client/hmi-client/src/services/workflow.ts

```diff
@@ -290,6 +290,51 @@ export const updateNodeStatus = (wf: Workflow, nodeId: string, status: OperatorS
 	node.status = status;
 };
 
+export const getNeighborNodes = (wf: Workflow, id: string) => {
+	const inputEdges = wf.edges.filter((edge) => edge.target === id);
+	const outputEdges = wf.edges.filter((edge) => edge.source === id);
+	const upstreamNodes = inputEdges
+		.map((edge) => wf.nodes.find((n) => n.id === edge.source))
+		.filter(Boolean) as WorkflowNode<any>[];
+	const downstreamNodes = outputEdges
+		.map((edge) => wf.nodes.find((n) => n.id === edge.target))
+		.filter(Boolean) as WorkflowNode<any>[];
+	return { upstreamNodes, downstreamNodes };
+};
```

Comment:
You should build some kind of node-cache (id => node) to avoid unnecessary `O(n^2)` lookups. 

Also, consider adding in a unit test case for this.

Also, now you need to account for tombstone logic.